server/embed/config_tracing.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package embed
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpgrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+	"go.uber.org/zap"
+)
+
+const maxSamplingRatePerMillion = 1000000
+
+func (e *Etcd) setupTracingExporter(ctx context.Context) (exporter tracesdk.SpanExporter, options []otelgrpc.Option, err error) {
+	exporter, err = otlp.NewExporter(ctx,
+		otlpgrpc.NewDriver(
+			otlpgrpc.WithEndpoint(e.cfg.ExperimentalDistributedTracingAddress),
+			otlpgrpc.WithInsecure(),
+		))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	res := resource.NewWithAttributes(
+		semconv.ServiceNameKey.String(e.cfg.ExperimentalDistributedTracingServiceName),
+	)
+
+	if resWithIDKey := determineResourceWithIDKey(e.cfg.ExperimentalDistributedTracingServiceInstanceID); resWithIDKey != nil {
+		// Merge resources into a new
+		// resource in case of duplicates.
+		res = resource.Merge(res, resWithIDKey)
+	}
+
+	sampler, err := determineSampler(e.cfg.ExperimentalDistributedTracingSamplingRatePerMillion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	options = append(options,
+		otelgrpc.WithPropagators(
+			propagation.NewCompositeTextMapPropagator(
+				propagation.TraceContext{},
+				propagation.Baggage{},
+			),
+		),
+		otelgrpc.WithTracerProvider(
+			tracesdk.NewTracerProvider(
+				tracesdk.WithBatcher(exporter),
+				tracesdk.WithResource(res),
+				tracesdk.WithSampler(tracesdk.ParentBased(sampler)),
+			),
+		),
+	)
+
+	e.cfg.logger.Debug(
+		"distributed tracing enabled",
+		zap.String("address", e.cfg.ExperimentalDistributedTracingAddress),
+		zap.String("service-name", e.cfg.ExperimentalDistributedTracingServiceName),
+		zap.String("service-instance-id", e.cfg.ExperimentalDistributedTracingServiceInstanceID),
+		zap.Int("sampling-rate", e.cfg.ExperimentalDistributedTracingSamplingRatePerMillion),
+	)
+
+	return exporter, options, err
+}
+
+func determineSampler(samplingRate int) (tracesdk.Sampler, error) {
+	sampler := tracesdk.NeverSample()
+	if samplingRate == 0 {
+		return sampler, nil
+	}
+	if samplingRate < 0 {
```

Comment:
This looks like flag value validation, please move it to embed.Config.Validate

Reply:
Good suggestion, for both this and the below check right?

Reply:
I think so. Best would be to have a separate configuration struct for tracing that is included in embed.Config. This way we can have a validate function in tracing package that is invoked by embed.Config.Validate (common pattern in K8s code).

Reply:
yeah agreed its bit different per configuration here. The cfg.setupLogging seems to be called in validate as well. But agreed we can do it down the road in similar way, especially since config is really growing and lacks testing IMO.